src/vision/status_quo/alan_finds_database_drops_hard.md

```diff
@@ -0,0 +1,104 @@
+
+
+## The problem
+
+Alan has been adding an extension to YouBuy that launches a singleton actor which interacts with a Sqlite database using the `sqlx` crate. The Sqlite database only permits a single active connection at a time, but this is not a problem, because the actor is a singleton, and so there only should be one at a time. He consults the documentation for `sqlx` and comes up with the following code to create a connection and do the query he needs:
+  
+```rust
+use sqlx::Connection;
+
+#[async_std::main]
+async fn main() -> Result<(), sqlx::Error> {
+    // Create a connection
+
+    let conn = SqliteConnection::connect("sqlite::memory:").await?;
+
+    // Make a simple query to return the given parameter
+    let row: (i64,) = sqlx::query_as("SELECT $1")
+        .bind(150_i64)
+        .fetch_one(&conn).await?;
+
+    assert_eq!(row.0, 150);
+
+    Ok(())
+}
+```
+
+Things seem to be working fairly well but sometimes when he refreshes the page he encounters a panic with the message "Cannot open a new connecton: connection is already open". He is flummoxed.
+
+
+## Searching for the Solution
+
+
+Alan tries to figure out what happened from the logs, but the only information he sees is that a new connection has been received. Alan turns to the documentation for the `sqlx` crate to see if there are flags that might enable extra instrumentation but he can't find any [sqlx::Connection::close](https://docs.rs/sqlx/0.5.1/sqlx/trait.Connection.html#required-methods).
+
+* He does find the [`close` method] which mentions "This method is not required for safe and consistent operation. However, it is recommended to call it instead of letting a connection drop as the database backend will be faster at cleaning up resources."
+* He adds a call to `close` into his code and it helps some but he is still able to reproduce the problem if he refreshes often enough. 
+* He adds a log statement right before calling `close` to see if it is working:
+
+```rust
+use sqlx::Connection;
+
+#[async_std::main]
+async fn do_the_thing() -> Result<(), sqlx::Error> {
+    // Create a connection
+    let conn = SqliteConnection::connect("sqlite::memory:").await?;
+
+    // Make a simple query to return the given parameter
+    let row: (i64,) = sqlx::query_as("SELECT $1")
+        .bind(150_i64)
+        .fetch_one(&conn).await?; // <----- if this await is cancelled, doesn't help
+
+    assert_eq!(row.0, 150);
+    
+    // he adds this:
+    log!("closing the connection");
+    conn.close();
+
+    Ok(())
+}
+```
+  
+* He observes that in the cases where he has the problem the log statement never executes.
+* He asks Barbara for help and she explains how `await` can be canceled and it will the destructors for things that are in scope.
```

Comment:
I wonder if we can link to something here that explains this-- googling brought up [this gist](https://gist.github.com/Matthias247/ffc0f189742abf6aa41a226fe07398a8) by @Matthias247, which could be good

Reply:
```suggestion
* He asks Barbara for help and she points him to [this gist](https://gist.github.com/Matthias247/ffc0f189742abf6aa41a226fe07398a8) that explains how `await` can be canceled and it will the destructors for things that are in scope.
```